src/eventer/eventer_SSL_fd_opset.c

```diff
@@ -215,69 +282,68 @@ save_dh_params(DH *p, const char *filename) {
   return;
 }
 
-static DH *dh1024_tmp = NULL, *dh2048_tmp = NULL;
-static const char *dh1024_file = NULL, *dh2048_file = NULL;
 static int
 generate_dh_params(eventer_t e, int mask, void *cl, struct timeval *now) {
   (void)e;
   (void)now;
-  int bits = (int)(intptr_t)cl;
+  struct stat sb;
+  char errstr[120];
+  unsigned int err;
+  int rv;
   if(mask != EVENTER_ASYNCH_WORK) return 0;
-  switch(bits) {
-  case 1024:
-    if(!dh1024_tmp) dh1024_tmp = load_dh_params(dh1024_file);
-    if(!dh1024_tmp) {
-      mtevL(mtev_notice, "Generating 1024 bit DH parameters.\n");
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-      dh1024_tmp = DH_generate_parameters(1024, 2, NULL, NULL);
-#else
-      int problems = 0;
-      DH *dh_tmp_trans = DH_new();
-      DH_generate_parameters_ex(dh_tmp_trans, 1024, 2, NULL);
-      if(DH_check(dh_tmp_trans, &problems) == 1 && problems == 0) {
-        dh1024_tmp = dh_tmp_trans;
-      } else {
-        DH_free(dh_tmp_trans);
+  ERR_clear_error();
+  struct dhparams_t *tgt = (struct dhparams_t *)cl;
+
+  ERR_clear_error();
+  if(tgt->file) {
+    memset(&sb, 0, sizeof(sb));
+    if(!tgt->params) {
+      while((rv = stat(tgt->file, &sb)) == -1 && errno == EINTR);
+      if(rv == 0 || errno != ENOENT) {
+        tgt->params = load_dh_params(tgt->file, tgt->bits);
       }
-#endif
-      mtevL(mtev_notice, "Finished generating 1024 bit DH parameters.\n");
-      if(dh1024_tmp) save_dh_params(dh1024_tmp, dh1024_file);
     }
-    break;
-  case 2048:
-    if(!dh2048_tmp) dh2048_tmp = load_dh_params(dh2048_file);
-    if(!dh2048_tmp) {
-      mtevL(mtev_notice, "Generating 2048 bit DH parameters.\n");
+    if(!tgt->params) {
+      mtevL(mtev_notice, "Generating %d bit DH parameters.\n", tgt->bits);
 #if OPENSSL_VERSION_NUMBER < 0x10100000L
-      dh2048_tmp = DH_generate_parameters(2048, 2, NULL, NULL);
+      tgt->params = DH_generate_parameters(tgt->bits, 2, NULL, NULL);
+      if(!tgt->params) {
+        while(0 != (err = ERR_get_error())) {
+          mtevL(eventer_err, " -> %s\n", ERR_error_string(err, errstr));
```

Comment:
here errstr is used, should also send the string length (120) so that it can be checked for overflow

Reply:
ERR_error_string doesn't take length, it expects the string to be long enough.  There is a newer API, but it would be worse (IMO) to use it b/c we'd need to ifdef to support older OpenSSLs.

Reply:
ok no need, just #define the 120